FEATURE - CONVERSIONS HISTORY LOGIC

Deivid117 opened the pull request:
Módulo Core
Data
- Se agregó la base de datos de room y el type converter para guardar local date time en room

Módulo History
Data
- Se agregaron los dao, entities y lógica del repository para obtener el historial de conversiones
Domain
- Se agregaron los modelos necesarios, repository y use case
Presentation
- Se cambió la interfaz de la pantalla de Conversion History, posteriormente se regresó a la interfaz original

[Tareas](https://trello.com/c/41WuSn5d)

Deivid117 committed:
FEATURE - History module logic

Se agregó el archivo para la creación de la bd ROOM así como también su inyección de dependencias.
De igual forma ya está implementada la lógica con los DAO y repository para obtener el historial de conversiones

Deivid117 committed:
Merge branch 'develop' of https://github.com/VarquLabs/DolarBlueApp into feature/conversion-history-logic

Deivid117 committed:
REFACTOR - Room database version

Se restauró la versión de la bd de room a la versión 1

Deivid117 committed:
REFACTOR - ConversionHistoryScreen update

Se restauró la pantalla al código original

DarioPapucho commented:
Tienes 53 cambios, entre ellos esta el merge que ya se realizo en la rama base de Elias y Yo.
Actualizaste tu rama develop y le hiciste rebase o merge de la rama trabajada a la nueva rama actualizada develop?
Quiza por eso es que te salen demasiados cambios

eliasmp07 commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/dao/ConversionHistoryDao.kt:
```diff
@@ -0,0 +1,13 @@
+interface ConversionHistoryDao {
+    @Transaction
+    @Query("SELECT * FROM current_exchange_rate_table")
+    fun getConversionsHistory(): List<ConversionsHistoryRelation>
```
Es mejor usar Flow en la base de datos Room(Si habra cambios en la base de datos)

Deivid117 committed:
REFACTOR - Conversion History logic

Se agregaron espacios entre elementos en los archivos y flows a los retornos de los repositorios, dao y use case de las conversiones

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/core/data/local/database/converters/DBTypeConverter.kt:
```diff
@@ -0,0 +1,17 @@
+class LocalDateTimeConverter {
+
+    @TypeConverter
+    fun toDate(dateString: String?): LocalDateTime? {
```
toLocalDateTime

juanjosevarqu replied:
?

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/dao/ConversionDao.kt:
```diff
@@ -1,5 +1,12 @@
+import androidx.room.Insert
+import com.varqulabs.dolarblue.history.data.local.database.entities.ConversionEntity
+
+// TODO @Deivid - mover el dao al módulo de calculadora
```
Podemos moverlo de una vez al modulo de calculadora porfa

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/domain/model/Conversion.kt:
```diff
@@ -1,6 +1,12 @@
 package com.varqulabs.dolarblue.history.domain.model
 
-// TODO @Deivid - posteriormente se modificará*/
+import java.time.LocalDateTime
```
En el proximo PR, junto con el Reemplazo de KSP, hagamos el reemplazo de Kotlin LocalDateTime porfa

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/dao/ConversionDao.kt:
```diff
@@ -1,5 +1,12 @@
+// TODO @Deivid - mover el dao al módulo de calculadora
+@Dao
 interface ConversionDao {
+    @Insert()
```
usemos el onConflict REPLACE plox

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/dao/ConversionHistoryDao.kt:
```diff
@@ -0,0 +1,14 @@
+interface ConversionHistoryDao {
+    @Transaction
+    @Query("SELECT * FROM current_exchange_rate_table")
+    fun getConversionsHistory(): Flow<List<ConversionsHistoryRelation>>
```
creemos una funcion auxiliar que no dependa del Flow, para operaciones que solo requieran ser llamadas una sola vez

eliasmp07 replied:
Si los habrá cambios constantes se le pondrá al nombre Flow al final y se usara flow, sin no abra cambios se manejara Listas dado que no abra cambios en la base de datos y no se estaba escuchando.

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/dao/CurrentExchangeRateDao.kt:
```diff
@@ -1,5 +1,13 @@
+import androidx.room.OnConflictStrategy
+import com.varqulabs.dolarblue.history.data.local.database.entities.CurrentExchangeRateEntity
+
+// TODO @Deivid - mover el dao al módulo de calculadora
```
.

juanjosevarqu commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/entities/ConversionEntity.kt:
```diff
@@ -1,6 +1,28 @@
+import com.varqulabs.dolarblue.history.domain.model.Conversion
+import java.time.LocalDateTime
+
+// TODO: @Deivid - mover el entity al módulo de calculadora
```
.

eliasmp07 commented on app/src/main/java/com/varqulabs/dolarblue/history/data/local/database/entities/relations/ConversionsHistoryRelation.kt:
```diff
@@ -0,0 +1,22 @@
+    val conversions: List<ConversionEntity>
+)
+
+fun ConversionsHistoryRelation.mapToModel() = ConversionsHistory(
```
Se movera a un carpeta llamada mappers en data, se creara un archivo que contenga los mappers de ConversionsHistory por ejemplo: ConversionsHistoryRelationMapper

eliasmp07 reviewed (commented):
Deje comentarios para revisar.

juanjosevarqu reviewed (changes requested):
Deje un par de comentarios

Deivid117 committed:
REFACTOR - updated files

Se movieron archivos de database del módulo de history al de calculator ya que ahí se usarán posteriormente.
Se agregó una fun más para el Dao conversion history para retornar flow o no.
Se renombraron archivos

juanjosevarqu reviewed (commented):
Falto el Renombre de la funcion de Parseador de Fecha

Deivid117 committed:
REFACTOR - type converter

se cambió el nombre del type converter para el LocalDateTime
